codec: a few suggestions

bluetech opened the pull request:
Please see the commits.

carllerche commented on tokio-codec/src/framed_read.rs:
```diff
@@ -211,7 +211,9 @@ where
             // got room for at least one byte to read to ensure that we don't
             // get a spurious 0 that looks like EOF
-            pinned.buffer.reserve(1);
+            if pinned.buffer.capacity() <= pinned.buffer.len() {
```
`reserve` ensures that `capacity - length >= reserve_amount`. This means that `reserve(1)` is a no-op when `!pinned.buffer.is_empty()`. The additional guard is not needed. See [source](https://github.com/tokio-rs/bytes/blob/b5d4f87e7e6b7d9636aa42d3d3bd63503579b0b6/src/bytes.rs#L2307-L2315).

I would request reverting that commit.

bluetech commented:
Right, got confused. I removed the misguided commit. I also changed the other commit accordingly, and made the example also reserve space for the next frame's header, which seems sensible.
